src/EventBuilder/Cecil/StaticEventTemplateInformation.cs

```diff
@@ -0,0 +1,158 @@
+            var type = ei.EventType.Resolve();
+            var invoke = type.Methods.First(x => x.Name == "Invoke");
+            if (invoke.Parameters.Count < 1) return null;
+
+            var param = invoke.Parameters.Count == 1 ? invoke.Parameters[0] : invoke.Parameters[1];
+            var ret = RenameBogusWinRTTypes(param.ParameterType.FullName);
+
+            var generic = ei.EventType as GenericInstanceType;
+            if (generic != null)
+            {
+                foreach (
+                    var kvp in
+                        type.GenericParameters.Zip(generic.GenericArguments, (name, actual) => new {name, actual}))
+                {
+                    var realType = GetRealTypeName(kvp.actual);
+
+                    ret = ret.Replace(kvp.name.FullName, realType);
+                }
+            }
+
+            // NB: Inner types in Mono.Cecil get reported as 'Foo/Bar'
+            return ret.Replace('/', '.');
+        }
+
+        private static string GetRealTypeName(TypeDefinition t)
+        {
+            if (t.GenericParameters.Count == 0) return RenameBogusWinRTTypes(t.FullName);
+
+            var ret = string.Format("{0}<{1}>",
+                RenameBogusWinRTTypes(t.Namespace + "." + t.Name),
+                string.Join(",", t.GenericParameters.Select(x => GetRealTypeName(x.Resolve()))));
+
+            // NB: Inner types in Mono.Cecil get reported as 'Foo/Bar'
+            return ret.Replace('/', '.');
+        }
+
+        private static string GetRealTypeName(TypeReference t)
+        {
+            var generic = t as GenericInstanceType;
+            if (generic == null) return RenameBogusWinRTTypes(t.FullName);
+
+            var ret = string.Format("{0}<{1}>",
+                RenameBogusWinRTTypes(generic.Namespace + "." + generic.Name),
+                string.Join(",", generic.GenericArguments.Select(x => GetRealTypeName(x))));
+
+            // NB: Inner types in Mono.Cecil get reported as 'Foo/Bar'
+            return ret.Replace('/', '.');
+        }
+
+        private static EventDefinition[] GetPublicEvents(TypeDefinition t)
+        {
+            return
+                t.Events
+
+                    .Where(x =>
+                    {
+                        return x.AddMethod.IsPublic && GetEventArgsTypeForEvent(x) != null;
+                    })
+                    .ToArray();
+        }
+
+        public static NamespaceInfo[] Create(AssemblyDefinition[] targetAssemblies)
+        {
+            var publicTypesWithEvents = targetAssemblies
+                .SelectMany(x => SafeTypes.GetSafeTypes(x))
+                .Where(x => x.IsPublic && !x.HasGenericParameters)
+                .Select(x => new {Type = x, Events = GetPublicEvents(x)})
+                .Where(x => x.Events.Length > 0)
+                .ToArray();
+
+            var garbageNamespaceList = new[]
+            {
+                "Windows.UI.Xaml.Data",
```

Comment:
Likewise, are these namespaces sensible to exclude?

Reply:
Same, I am cleaning all of the, seemingly, non-relevant stuff out of here.